src/algebra/Module/basic.lean

```diff
@@ -0,0 +1,55 @@
+/-
+Copyright (c) 2019 Robert A. Spencer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert A. Spencer
+-/
+import algebra.module
+import algebra.punit_instances
+import category_theory.types
+import linear_algebra.basic
+open category_theory
+
+universe u
+
+variables (R : Type u) [ring R]
+
+/-- The category of R-modules and their morphisms. -/
+structure Module :=
+  (carrier : Type u)
+  [is_add_comm_group : add_comm_group carrier]
+  [is_module : module R carrier]
+
+attribute [instance] Module.is_add_comm_group Module.is_module
+
+namespace Module
+
+instance : has_coe_to_sort (Module R) :=
+  { S := Type u, coe := Module.carrier }
+
+def of (X : Type u) [add_comm_group X] [module R X] : Module R := ⟨R, X⟩
+
+instance : has_zero (Module R) := ⟨ of R punit ⟩
+
+variables (M N U : Module R)
+
+instance : category (Module R) := {
+  hom := λ M N, M →ₗ[R] N,
+  id := λ M, 1,
+  comp := λ A B C f g, g.comp f
+}
+
+@[simp] lemma id_apply (m : M) : (𝟙 M : M → M) m = m := rfl
+
+@[simp] lemma module_hom_comp (f : M ⟶ N) (g : N ⟶ U) :
+  ((f ≫ g) : M → U) = g ∘ f := rfl
+
+@[extensionality] lemma hom_ext  {f g : M ⟶ N} : (f : M → N) = g → f = g :=
+  λ w, linear_map.ext (function.funext_iff.1 w)
```

Comment:
No indentation here

Reply:
Also, I suppose the statement should match the one from `concrete_category`, namely `(∀ x : X, f x = g x) → f = g`.

Reply:
Oh, now I see this was already discussed. I would suggest just doing the same thing as `bundled` unless you have some specific reason not to.

Reply:
Is the one from bundled not just this composed with funext?

Reply:
Yes, that's true. I'm just saying that this same decision to include funext or not was already made in `bundled`, so a good default course of action is to assume that that decision was made correctly. If it's right to not include funext here then probably we should also go and change the one in `bundled`, unless there is something special about modules, which there doesn't seem to be.

Same argument applies to the fields of the structure `Module`.

(Basically the reason we don't just use `bundled` here is that we need to provide two different bits of structure, and `bundled` only lets us give one.)

Reply:
Pardon me for maybe being daft.  Would this suffice?
```lean
@[extensionality] lemma hom_ext {f g : M ⟶ N} : (∀ m : M, f m = g m) → f = g := linear_map.ext_iff.2
```

Reply:
As it turns out, this extensionality lemma is unnecessary, as `linear_map.ext` kicks in automatically. I've removed it.